[BE] refactor : 히어릿 탐색 테스트 작성 및 리팩토링

Byesol opened the pull request:
## 📌 변경 내용 & 이유

### 1. 탐색 점수 갱신 로직 분리 (`ExploreScoreRefresher` 도입)

**[기존 문제]**
기존 `AbstractExploreScoreProcessor`는 점수 초기화(계산 및 저장)와 조회(결과 반환)라는 두 가지 책임을 모두 가지고 있었습니다. 이로 인해 두 로직이 강하게 결합되어 있어, 점수 계산 정책을 변경하거나 비동기 처리 등을 도입하기 어려운 구조였습니다. 또한, 초기화와 조회를 분리하여 테스트하기 어려웠습니다.

**[개선 내용]**
점수 초기화(갱신) 책임만을 전담하는 `ExploreScoreRefresher` 클래스를 새로 도입했습니다.
- **`ExploreScoreRefresher`**: `cursorId`가 0일 때 (첫 페이지 요청 시) 점수 계산 및 저장을 담당합니다.
- **`ExploreScoreProcessor`**: 점수판을 기반으로 데이터를 조회하고 API 응답을 조립하는 책임만 남게 되었습니다.
- **`HearitExploreService`**: 두 컴포넌트를 조합하여 전체적인 흐름을 제어합니다.

이 분리의 핵심적인 이유는, 첫 페이지 요청(cursorId == 0) 시에만 점수를 갱신한다'는 규칙이 단순한 내부 구현 로직이 아니라, 사용자 행동과 직접적으로 관련된 중요한 '유스케이스(Use Case)'**이기 때문입니다. 이러한 비즈니스 시나리오는 하위 컴포넌트 내부에 감춰지는 게 아니라 서비스 계층의 흐름(Flow)에서 명확하게 읽혀야 한다고 생각했습니다.

그리고 현재 구조에서는 **점수 갱신 -> 점수판 조회**라는 작업의 순서가 보장되어야 합니다. 그러나 processor 내부에서 메서드를 통해 그 순서를 보장하는 것보단  ExploreScoreRefresher를 별도 클래스로 분리하고 서비스 계층에서 명시적으로 호출하면서, 유스케이스의 시나리오 순서를 서비스 코드가 직접 보장하도록 구조가 더 좋다고 생각합니다.

 과거 다른 미션(결제 미션)에서도 서비스 코드에서 순서가 중요했던 것처럼, 비즈니스 시나리오의 흐름과 순서를 서비스 계층에서 명확하게 드러내고 보장하는 것이 좋다고 생각합니다.

결과적으로 각 클래스는 단일 책임 원칙(SRP)을 따르게 되었고 만일 나중에 점수 갱신 정책(예: 비동기 처리, 캐싱, 스케줄링)을 변경하더라도 ExploreScoreRefresher만 수정하면 되는 걸로 바뀌었습니다.

### 2. 테스트 개선

기존 테스트 코드는 외부 요인(랜덤 값, 시간)에 의존적이어서 불안정하고, 테스트 간 격리가 어려워 함께 실행 시 실패하는 문제가 있었습니다.

**가. 랜덤(Random) 값 제어**
- **문제**: `RandomScoreFactor`가 `new Random()`에 직접 의존하여 테스트 시 예측 불가능한 값을 생성했습니다.
- **해결**: 랜덤 값 생성을 `RandomNumberGenerator` 인터페이스로 추상화하고, 테스트에서는 항상 고정된 값을 반환하는 구현체를 주입하여 랜덤 요소를 완벽하게 제어하도록 변경했습니다.

**나. 시간(Time) 제어 및 JPA Auditing 문제 해결**
- **문제**:
    1. `@EnableJpaAuditing` 기능이 활성화된 컨텍스트에서는 수동으로 설정한 시간이 DB 저장 시점에 **현재 시간으로 덮어써지는 문제**를 발견했습니다.
    2. 이 문제를 해결하기 위해 Auditing 설정을 켜고 끄는 두 개의 테스트 그룹으로 분리하자, 스프링의 **'테스트 컨텍스트 캐싱'** 문제로 인해 테스트 실행 순서에 따라 `NoSuchBeanDefinitionException` 오류가 발생하는 등 테스트 격리에 실패했습니다.

- **해결**:
    1. Auditing 기능이 시간을 가져오는 `DateTimeProvider`를 테스트에서 제어할 수 있도록 **`TestClock`** 이라는 커스텀 구현체를 도입했습니다.
    2. 이를 통해 **모든 테스트에서 Auditing 설정을 안전하게 켤 수 있는 단일화된 테스트 환경을 구축**했습니다. 컨텍스트 캐싱 문제가 원천적으로 해결되었습니다.
    3. 이제 테스트 코드에서 `TestClock.freezeAt(원하는_시간)`을 호출하기만 하면, JPA Auditing이 해당 시간을 `@CreatedDate`에 자동으로 주입해 줍니다. 이로써 안정적으로 시간 데이터를 제어할 수 있게 되었습니다.

### 3. 테스트 커버리지 강화

위 구조 변경을 바탕으로, 각 컴포넌트의 역할과 책임을 명확하게 검증하는 단위/통합 테스트를 추가했습니다.
- `BookmarkScoreFactorTest`, `RecencyScoreFactorTest`, `RandomScoreFactorTest` 등 점수 요인별 단위 테스트
- `ExploreScoreCalculatorTest`, `ExploreScoreRefresherTest` 등 핵심 로직 통합 테스트
- `MemberExploreScoreProcessorTest`, `GuestExploreScoreProcessorTest` 등 프로세서별 테스트
- 개선된 `HearitExploreServiceTest`

## 🧪 테스트 방법 (선택)
- `com.onair.hearit.app.application.explore.HearitExploreServiceTest`
- `com.onair.hearit.app.application.explore.ExploreScoreCalculatorTest`
- `com.onair.hearit.app.application.explore.ExploreScoreRefresherTest`
- 각 `scorefactor` 패키지 하위의 테스트 클래스들

## 📢 논의하고 싶은 내용
**1. `TestClock` 적용**
시간에 의존적인 로직을 테스트하는 새로운  방식으로 `TestClock`을 도입했습니다.  이에 대한 의견이 궁금합니다.

**2. `ExploreScoreRefresher` 네이밍**
초기 제안에서는 점수판의 라이프사이클을 관리한다는 의미에서 `ExploreScoreSessionManager`라는 이름도 고려했으나, 최종적으로는 '점수를 갱신한다'는 핵심 책임을 명확히 하고자 `ExploreScoreRefresher`로 명명했습니다. 네이밍에 대한 다른 좋은 의견이 있다면 말씀해주세요

3. 문제 상황: @DataJpaTest에서의 의존성 주입의 한계

@DataJpaTest는 JPA Repsitory처럼 데이터 계층과 관련된 빈(Bean)들만 자동으로 등록해 줍니다. ExploreScoreCalculator와 같은 @Component나 @Service는 스캔 대상이 아니므로, 단순히 @Autowired만 사용해서는 NoSuchBeanDefinitionException 오류가 발생합니다. 즉, @DataJpaTest 환경에서는 완전한 자동 의존성 주입이 불가능합니다.

-  시도했던 대안들과 문제점

이 문제를 해결하기 위해 두 가지 방식을 시도했습니다.

@Import({ComponentA.class, ComponentB.class, ...}) 방식:

문제점: 필요한 컴포넌트를 일일이 명시해야 해서 설정이 밑도 끝도 없이 길어졌습니다.

@BeforeEach에서 수동 생성(new Service(...)) 방식:

문제점: 테스트 코드의 준비(given) 단계가 복잡해지고, DI 컨테이너의 이점을 포기하게 되어 객체 간의 의존성이 변경될 때마다 테스트 코드를 직접 수정해야 하는 번거로움이 있었습니다.

다. 해결책: @TestConfiguration의 활용

최종적으로 @TestConfiguration을 활용하는 방식으로 해결했습니다.

 TestConfig 도입: ScoreFactor처럼 여러 테스트에서 공통으로 필요한 구현나 설정을 TestConfig에 @Bean으로 등록했습니다. 이제 다른 테스트에서는 @Import(TestConfig.class)   공통 설정을 가져와 사용할 수 있게 했는데 이에 대해 어떻게생각하시는지 궁금합니다. 


## 🧩 관련 이슈
close #588

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- 신규 기능
  - 탐색 피드가 첫 요청(cursorId == 0) 시 점수 데이터를 초기화·삽입해 최신 결과 제공.
  - 탐색 점수 초기화 컴포넌트와 랜덤 숫자 생성 인터페이스/기본 구현 추가.
- 리팩터
  - 점수 계산 흐름을 초기화 중심으로 재구성하고 일부 API 호출을 분리·명확화(메서드/의존성 교체).
- 테스트
  - 점수 계산·초기화·각 점수 팩터·프로세서에 대한 단·통합 테스트 대폭 추가.
- 잡무
  - 테스트용 시간 고정 유틸리티와 DB 헬퍼 확장(시간 지정 삽입, 탐색 점수 삽입 메서드명 변경) 추가.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Byesol committed:
[BE] refactor: 랜덤 점수 생성 책임 분리 및 인터페이스 도입 #588

- `RandomScoreFactor` 클래스에서 `Random`의 직접적인 의존성 제거
- 랜덤 값 생성을 담당하는 `RandomNumberGenerator` 인터페이스 추가
- `RandomNumberGenerator` 구현체로 `DefaultRandomNumberGenerator` 클래스 생성
- DI를 위해 `@RequiredArgsConstructor` 추가

Byesol committed:
[BE] refactor: explore score 갱신 책임 ExploreScoreRefresher로 분리 #588

- 탐색 점수 갱신 로직을 `ExploreScoreRefresher` 클래스로 이동
- `getExploreHearitsResponse`에 userUuid 매개변수 추가
- `resolveUserUuid` 메서드 인터페이스 추가 및 구현체 수정

Byesol committed:
[BE] refactor: 점수 팩터별 테스트 케이스 추가 및 TestConfig 구성 #588

- `RandomScoreFactorTest` 테스트 클래스 추가
- `BookmarkScoreFactorTest` 테스트 클래스 추가
- `RecencyScoreFactorTest` 테스트 클래스 추가
- 테스트에 필요한 `TestConfig` 구성 및 스프링 설정 추가

Byesol committed:
[BE] refactor: ExploreScore 관련 테스트 케이스 추가 및 기존 테스트 구조 개선 #588

- `MemberExploreScoreProcessorTest` 테스트 클래스 추가
- `ExploreScoreCalculatorTest` 테스트 클래스 추가
- `ExploreScoreRefresherTest` 테스트 클래스 추가
- `HearitExploreServiceTest`에 ExploreScore 관련 검증 시나리오 추가

Byesol committed:
[BE] refactor: 테스트 시나리오용 DateTimeProvider 및 관련 설정 추가 #588

- JPA Auditing 테스트를 위한 `TestClock` 클래스 추가
- `TestConfig` 클래스 추가
- `TestJpaAuditingConfig`에 `testDateTimeProvider` Bean 설정 및 적용

Byesol committed:
[BE] refactor: ExploreScoreCalculatorTest에 TestClock 활용 테스트 개선 #588

- `TestClock`을 활용해 테스트 간 시간 의존성을 제거하고 명시적으로 설정
- `TestJpaAuditingConfig` 클래스 임포트 추가
- `tearDown` 메서드를 통해 테스트 종료 시 `TestClock` 초기화
- `createHearitWithTime` 메서드 삭제 및 `createHearit` 메서드로 통합
- 테스트 시나리오 내 생성 시간 설정 로직 명시적으로 변경

Byesol committed:
[BE] refactor: RecencyScoreFactorTest에 TestClock 활용 및 테스트 개선 #588

- `TestClock`을 활용한 테스트 시나리오 시간 의존성 제거 및 명시적 설정 추가
- `DbHelper`, `TestConfig`, `TestJpaAuditingConfig` 임포트 추가

Byesol committed:
[BE] refactor: 테스트 코드에서 BASE_TIME 제거 및 TestClock 활용 #588

- `BASE_TIME` 제거 후 `LocalDateTime.now()`를 활용하여 테스트 코드 간소화
- `TestClock`을 사용해 시간 의존성을 제거하고 테스트의 명시성을 개선
- 중복된 시간 생성 로직을 통합하고 간결하게 정리

Byesol committed:
[BE] refactor: 불필요한 Hearit 생성자 제거 및 코드 정리 #588

- 중복된 Hearit 생성자 삭제로 코드 간소화
- 사용되지 않는 생성자 제거를 통해 유지 관리성 개선

Byesol committed:
[BE] refactor: 테스트 설정 클래스 이름 변경 및 관련 참조 수정 #588

- `TestConfig`를 `ExploreScoreTestConfig`로 이름 변경
- 테스트 클래스 전반에서 변경된 설정 클래스 이름 참조 수정

Byesol committed:
[BE] refactor: TestClock 주석 제거 및 메서드 정리 #588

Byesol committed:
[BE] refactor: 테스트 설정 클래스 TestConfig 삭제 #588

gabean13 commented on backend/app/src/main/java/com/onair/hearit/explore/application/HearitExploreService.java:
```diff
@@ -14,13 +14,17 @@
 
     public CursorResponseV2<ExploredHearitResponse> getExploredHearits(UserInfo userInfo,
                                                                        CursorRequest cursorRequest) {
         ExploreScoreProcessor exploreScoreProcessor = getExploreScoreProcessor(userInfo);
+        String userUuid = exploreScoreProcessor.resolveUserUuid(userInfo);
+        exploreScoreRefresher.refreshIfNeeded(cursorRequest.cursorId(), userUuid, userInfo.getUserType());
```
```suggestion
        ExploreScoreProcessor exploreScoreProcessor = getExploreScoreProcessor(userInfo);
        exploreScoreProcessor.refreshIfNeeded(cursorRequest.cursorId(), userInfo);
```

refresher라는 클래스를 추가하기 보다 이렇게 하는거 어떤가요??
왜냐면 지금 exploreScoreProcessor에서 resolverUserUuid를 받아서 -> 다시 exploreScoreRefresher에 점수를 갱신하고 -> 다시 exploreScoreProcessor에서 응답을 계산하잖아요.

이렇게 중간에 exploreScoreProcessor가 중간에 낀 이유가 결국 member, guest랑 다르기때문인데 지금 exploreScoreRefresher는 그걸 모르게 하고 싶으니까 그렇게 된 것 같아요. 
차라리 guest, member를 아는 애들이 refresh하는 책임도 가지면 중간에 resolverUserUuid를 service에서 호출할 필요도 없고 흐름도 더 이해가 잘될 것 같아용

대신 벡터가 말한 외부에서 갱신하는걸 알아야한다는건 동의하기때문에 public으로 메서드를 선언해서 호출해주면 되지 않을까요?

yuyoungrhee replied:
그럼 결국 ExploreScoreProcessor라는 클래스에 여러 책임이 혼재되게 된다는 점은 동일해지는 거 같음요
테스트할 때 ExploreScoreProcessor에 초기화로직까지 테스트해야된다는 점요

Byesol replied:
저도 가콩이 말한 문제  때문에 고민이 생겨서 제안서 하나 더 만드려다가 말았었는데 processor로부터 아예 UserInfo를 받아 userUuid를 반환하는 클래스를 또 따로 분리하는 건 어떰요? 

결국 나중에 점수 갱신 로직은 스케줄러든, 어떤 식으로든 리팩토링 될 수 있을 것 같은데 같은데 이 경우 exploreScoreProcessor에 초기화와 조회 두 가지 책임이 혼재되어 있어 유지 보수가 힘들 것 같아서요

Byesol replied:
논의 후 processor가 refresher를 가지도록 변경했습니다
https://github.com/woowacourse-teams/2025-hEARit/pull/615/commits/a9e2fb92d594883abb862010f4a39e69bd5c60ff

gabean13 commented on backend/core/src/testFixtures/java/com/onair/hearit/core/fixture/TestConfig.java:
```diff
@@ -0,0 +1,7 @@
+
+import org.springframework.boot.test.context.TestConfiguration;
+
+@TestConfiguration
+public class TestConfig {
+}
```
이거는 왜 잇는거에염?

Byesol replied:
필요없어져서 제거했어욥

gabean13 commented on backend/app/src/test/java/com/onair/hearit/explore/application/HearitExploreServiceTest.java:
```diff
@@ -51,120 +52,170 @@ class HearitExploreServiceTest {
+    void tearDown() {
+        TestClock.unfreeze();
+    }
+
```
테스트 대면으로 설명해주세용 넘 어려비

gabean13 reviewed (changes requested):
벡터 테스트 짜느라 고생했어욧

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/scorefactor/RecencyScoreFactorTest.java:
```diff
@@ -0,0 +1,82 @@
+        TestClock.freezeAt(LocalDateTime.now().minusDays(4));
+        Hearit medium = dbHelper.insertHearit(createHearit(category));
+        TestClock.freezeAt(LocalDateTime.now().minusDays(60));
+        Hearit old = dbHelper.insertHearit(createHearit(category));
```
TestClock.freezeAt()을 호출하고 dbHelper.insert()를 하는 방식은 JPA Auditing의 내부 동작을 알아야만 이해할 수 있겠네요.
해당 프로젝트는 DateTimeProvider로 시간을 제어하고 있고, Jpa Aduting을 사용하기 때문에 db insert 시점에서 시간이 같이 결정되는구나! << 이걸 이해하기 어렵지 않을까요

DbHelper에 생성 시간을 직접 지정하여 엔티티를 삽입하는 헬퍼 메서드를 만드는 건 어떤가요

```java
 /**
     * 지정된 생성 시간에 Hearit 엔티티를 저장합니다.
     * 내부적으로 TestClock을 사용하여 시간을 고정한 후, 안전하게 해제합니다.
     */
    public Hearit insertHearitAt(Hearit hearit, LocalDateTime creationTime) {
        try {
            TestClock.freezeAt(creationTime);
            return insertHearit(hearit);
        } finally {
            // 이 메서드 내에서 시간 제어가 끝나므로 바로 unfreeze하여 다른 테스트에 영향을 주지 않도록 합니다.
            TestClock.unfreeze();
        }
    }
```

gabean13 replied:
멍구 진짜 핵심을 찔럿다

Byesol replied:
진짜 좋은데요? 그렇게 수정했어요. https://github.com/woowacourse-teams/2025-hEARit/pull/615/commits/cf691b25e194b3cf9351e85874c8b3c663502800

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/scorefactor/RecencyScoreFactorTest.java:
```diff
@@ -0,0 +1,82 @@
+                () -> assertThat(scores).hasSize(hearits.size()),
+                () -> assertThat(scores.get(latest.getId())).isEqualTo(20.0),
+                () -> assertThat(scores.get(medium.getId())).isEqualTo(18.0),
+                () -> assertThat(scores.get(old.getId())).isEqualTo(0.0)
```
딴얘기인데 상수명이 1일당 2포인트인 줄 알았는데 2일당 1포인트더라고요
 
 차라리 1일당 0.5포인트로 하고 헷갈리지 않도록 리팩터링하는 거 어떨까요 (어차피 score가 최종 계산될 때 Double임)
 
 그리고 상수명도좀헷갈려서 제안해봅니다요 
기존: DAYS_PER_POINT
요청: POINT_LOSS_PER_DAY

Byesol replied:
전체적으로 리팩토링했어요!
https://github.com/woowacourse-teams/2025-hEARit/pull/615/commits/906e72ab261908b98fb6f6b1f0e915551bffd4ad

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/scorefactor/RecencyScoreFactorTest.java:
```diff
@@ -0,0 +1,82 @@
+        TestClock.unfreeze();
+    }
+
+    @DisplayName("최근 히어릿일수록 더 높은 최신성 점수를 부여한다")
```
테스트만 보고는 RecencyScoreFactor가 어떻게 동작하는지 알기 어려운 것 같아요.

"당일 생성된 히어릿은 최대 점수(20점)를 받는다"
"생성 후 4일이 지난 히어릿은 2점 감소한 18점을 받는다 (하루 0.5점 감소)"
"생성 후 40일이 지난 히어릿은 최소 점수(0점)를 받는다"
"생성 후 40일이 초과한 히어릿도 최소 점수(0점) 이하로 내려가지 않는다"

등등 이런식으로 테스트해주면 더 좋지 않을까요

Byesol replied:
테스트케이스 추가했습니다!
https://github.com/woowacourse-teams/2025-hEARit/pull/615/commits/ac314cd06d88c659292dd0da2bb5028f4b209fb3

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/scorefactor/RandomScoreFactorTest.java:
```diff
@@ -0,0 +1,56 @@
+    @BeforeEach
+    void setup() {
+        randomScoreFactor = new RandomScoreFactor(() -> 0.1d);
+    }
```
setup으로 빼둔 이유가 있을까요? 아래에서 왜 1점이지? 하고 보니까 setup에서 0.1d로 설정해주고 있더라고요.
테스트 하나만 읽어도 그 테스트를 이해할 수 있어야된다고 생각합니다~

Byesol replied:
테스트 하나만 읽어도 이해할 수 있게 하려면

각각의 테스트마다 임의의 new RandomGenerator 를 만들어 new RandomScoreFactor(가짜_Generator) -> new ExploreScoreCalculator(..., 가짜_Factor) -> ... 와 같이 의존성의 최하단부터 최상단 서비스까지 주입해야돼서  

@MockBean 을 통해 RandomNumberGenerator 만 바꿔치고

` given(randomNumberGenerator.nextDouble()).willReturn(0.1d);`

위 예시처럼 각 테스트 내부에서 가짜 객체의 행동을 원하는 대로 정의했습니다

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/scorefactor/RandomScoreFactorTest.java:
```diff
@@ -0,0 +1,56 @@
+        for (int i = 1; i <= 5; i++) {
+            Hearit hearit = dbHelper.insertHearit(TestFixture.createFixedHearitWith(category));
+            hearits.add(hearit);
+        }
```
` 외부에서 주입된 고정된 '랜덤' 값(0.1)을 가지고 점수(1.0)가 올바르게 계산되는가?` 가 이 테스트의 목적이라면 5개까지 동일한 테스트를 반복할 필요는 없을 것 같아요. 어차피 랜덤을 제어하고 있는데 1~10범위안에 있는지, 값이 1인지 5번 검증할 필요는 없어 보여요

Byesol replied:
랜덤 생성기에 대한 테스트를 '랜덤 제어할 때', '랜덤 제어하지 않을 때' 로 나눠 제어 안 할때는 범위 내 값인지 5번 검증하고 제어하지 않을 때는 굳이 여러번 검증하지 않도록 했습니다.
https://github.com/woowacourse-teams/2025-hEARit/pull/615/commits/94500b73824b3b07d6c85f011906981280454723

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/app/explore/application/scorefactor/BookmarkScoreFactorTest.java:
```diff
@@ -0,0 +1,111 @@
+@Import({DbHelper.class, TestJpaAuditingConfig.class})
+@ActiveProfiles("fake-test")
+@AutoConfigureTestDatabase(replace = Replace.NONE)
+class BookmarkScoreFactorTest {
```
사소하지만
- UserType이 MEMBER가 아닌 경우 지원하지 않는다.
- 존재하지 않는 사용자(member) UUID로 계산을 시도하면 NotFoundException 예외가 발생한다.

등등도 필요하지 않을까요?

Byesol replied:
테스트 케이스 추가했어요
https://github.com/woowacourse-teams/2025-hEARit/pull/615/commits/2434b6bc91ae37792ea63ad2bb077a2a002ad26c

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/scoreprocessor/GuestExploreScoreProcessorTest.java:
```diff
@@ -0,0 +1,122 @@
+    private HearitKeywordRepository hearitKeywordRepository;
+
+    @Autowired
+    private ExploreScoreRefresher exploreScoreRefresher;
```
Processor의 관심사에서 Refresher를 제거하기 위해 클래스를 분리했는데 테스트에서 Refresher가 사용됨으로써 오히려 이해가 어려워지는 것 같습니다. 테스트를 온전히 이해하려면 다시 Refresher에 가서 뭐하는 놈인지 보고 와야 됐거든요.

해당 의존성을 제거하고서도 테스트가 이루어질 수 있을 것 같은데요, 아래 코멘트로 남겨두겠습니다.

Byesol replied:
refresher를 processor 내부로 옮김에 따라 전체적으로 수정했습니다!

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/scoreprocessor/GuestExploreScoreProcessorTest.java:
```diff
@@ -0,0 +1,122 @@
+        dbHelper.insertHearitKeyword(new HearitKeyword(hearit2, keyword));
+        dbHelper.insertHearitKeyword(new HearitKeyword(hearit3, keyword));
+
+        exploreScoreRefresher.refreshIfNeeded(0L, GUEST_ID, UserType.GUEST);
```
when이 검증하고자 하는 바에 의하면 이부분 쫙 없애도 테스트가 성공하는 것 같습니다

Byesol replied:
원래 AbstractProcessor 에 prepareKeywordsMap 가 있기 때문에 keyword를 응답값에 알맞게 매핑하는지를 when 에서 검증해야하는데 빠져있었어서 이를 새로운 코드의 when 절에는 추가했습니다!

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/scoreprocessor/GuestExploreScoreProcessorTest.java:
```diff
@@ -0,0 +1,122 @@
+        dbHelper.insertHearitKeyword(new HearitKeyword(hearit2, keyword));
+        dbHelper.insertHearitKeyword(new HearitKeyword(hearit3, keyword));
+
+        exploreScoreRefresher.refreshIfNeeded(0L, GUEST_ID, UserType.GUEST);
```
Refresher에 의존하지 않고

``` java
dbHelper.insertExploreScore(new ExploreScore(guestId, hearit1.getId(), 30.0, 1L));
dbHelper.insertExploreScore(new ExploreScore(guestId, hearit2.getId(), 20.0, 2L));
dbHelper.insertExploreScore(new ExploreScore(guestId, hearit3.getId(), 10.0, 3L));
```

이런식으로 exploreScore를 직접 넣어주는 게 더 테스트 이해에 좋지 않을까요? 초기화&점수계산 로직이 빠지면서 Processor는 온전히 탐색데이터를 조회하고 가공하는 역할을 하는만큼 그전까지는 데이터가 raw하게 정의되어있으면 가독성이 좋을 것 같습니다.

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/scoreprocessor/GuestExploreScoreProcessorTest.java:
```diff
@@ -0,0 +1,122 @@
+@AutoConfigureTestDatabase(replace = Replace.NONE)
+class GuestExploreScoreProcessorTest {
+
+    private static final String GUEST_ID = UUID.randomUUID().toString();
```
마찬가지로 각 독립된 테스트만으로는 이해하기 어려워지는 상수 같습니다. 
차라리 각 테스트 given에 
`String guestId = UUID.randomUUID().toString();`
이라고 정의해주는 게 읽기 편하고 독립된 테스트 환경을 만들어줄 수 있을 것 같습니다.

Byesol replied:
각각의 테스트 내에서 정의하도록 수정했습니다!

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/scoreprocessor/MemberExploreScoreProcessorTest.java:
```diff
@@ -0,0 +1,134 @@
+        dbHelper.insertHearitKeyword(new HearitKeyword(hearit3, keyword));
+        dbHelper.insertBookmark(TestFixture.createFixedBookmark(member, hearit1));
+
+        exploreScoreRefresher.refreshIfNeeded(0L, member.getUuid(), UserType.MEMBER);
```
GuestExploreScoreTest와 마찬가지로
이 부분은 해당 테스트에 필요 없는 코드입니다.

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/scoreprocessor/MemberExploreScoreProcessorTest.java:
```diff
@@ -0,0 +1,134 @@
+    private HearitKeywordRepository hearitKeywordRepository;
+
+    @Autowired
+    private ExploreScoreRefresher exploreScoreRefresher;
```
여기도 Refresher 의존을 없애는 방향으로 테스트해주세요!

Byesol replied:
모든 Processor가 Refresher를 들고 있게 됨에 따라 의존하게 되었습니다.

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/HearitExploreServiceTest.java:
```diff
@@ -51,120 +52,170 @@ class HearitExploreServiceTest {
+        // -- 북마크 이력용 Hearit들 --
+        TestClock.freezeAt(LocalDateTime.now());
+        // 최신성 20, 카테고리 22.5
+        Hearit hearit1 = dbHelper.insertHearit(createHearit(category1));
```
이부분도 dbHelper 캡슐화를 통해 가독성을 해결해볼 수 있겠네요

Byesol replied:
해당 부분은 dbHelper를 통해 개선되었습니다!

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/ExploreScoreRefresherTest.java:
```diff
@@ -0,0 +1,106 @@
+        exploreScoreRefresher.refreshIfNeeded(cursorId, GUEST_UUID, UserType.GUEST);
+
+        // then
+        List<ExploredHearitProjection> projections = exploredHearitQueryRepository
```
Refresher의 테스트가 QueryRepository의 로직에 의존하게 되는 문제가 있습니다. explore_score 테이블의 상태를 직접 검증하는 편이 좋을 거 같은데 이정도는 그냥 흐린눈할까요?

Byesol replied:
refresher, processor 모두 각각의 테스트들에서 queryRepository의 로직에 의존하지 않도록 수정했습니다.

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/ExploreScoreRefresherTest.java:
```diff
@@ -0,0 +1,106 @@
+@AutoConfigureTestDatabase(replace = Replace.NONE)
+class ExploreScoreRefresherTest {
+
+    private static final String GUEST_UUID = UUID.randomUUID().toString();
```
여기도 마찬가지입니다~
각 테스트 메서드 내에서 지역 변수로 선언해서 테스트 독립성을 보장합시두

Byesol replied:
지역 변수로 옮겼습니당

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/ExploreScoreRefresherTest.java:
```diff
@@ -0,0 +1,106 @@
+        // given
+        long cursorId = 0L;
+        Category category = dbHelper.insertCategory(TestFixture.createFixedCategory());
+        TestFixture.createFixedHearitWith(category);
```
사용하고있지 않은 코드입니다. 삭제해도 잘 동작합니다

Byesol replied:
해당부분 포함해 ExploreRefresherTest의 경우 전체적으로 수정되었습니다!

yuyoungrhee reviewed (changes requested):
리뷰가 좀 많아졌네요 미안합니다.
테스트코드 품질을 수호했습니다...ㅋ

JO-eusan commented on backend/app/src/main/java/com/onair/hearit/explore/application/scorefactor/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,6 @@
+
+public interface RandomNumberGenerator {
+
+    double nextDouble();
```
nextDouble() 보다는 generate()와 같은 숫자를 생성하는 의미가 담기면 좋을 것 같아요!

Byesol replied:
논의 후 아래와 같이 바꿨습니다!
https://github.com/woowacourse-teams/2025-hEARit/pull/615/commits/73bfe49c29f8a697f4f92c956e737b6808812bb6

JO-eusan commented on backend/core/src/testFixtures/java/com/onair/hearit/common/TestClock.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Override
+    public Optional<TemporalAccessor> getNow() {
+        LocalDateTime now = (fixedTime != null) ? fixedTime : LocalDateTime.now();
```
`LocalDateTime now = (fixedTime == null) ? LocalDateTime.now() : fixedTime`
이렇게 긍정문을 앞에 두는게 어떤가요?

Byesol replied:
https://github.com/woowacourse-teams/2025-hEARit/pull/615/commits/2f882083ef8577fec12f03d08ccd5e0953d042bf

반영완료!

JO-eusan reviewed (changes requested):
벡터 고생하셨어요! 코드 스타일 몇가지 커멘트 남겼어요!

Byesol committed:
[BE] refactor: ExploreScoreProcessor 가 Refresher 를 가지도록 리팩토링 #588

- `resolveUserUuid` 메서드 삭제 및 관련 호출 코드 제거
- `getExploreHearitsResponse` 메서드의 `userUuid` 파라미터 제거로 메서드 시그니처 단순화
- `ExploreScoreRefresher` 주입 및 `AbstractExploreScoreProcessor`로 로직 이동
- `MemberExploreScoreProcessor` 및 `GuestExploreScoreProcessor` 리팩토링 및 생성자 업데이트

Byesol committed:
[BE] refactor: RandomScoreFactorTest db 의존 제거 후 Hearit 생성자 추가 #588

Byesol committed:
[BE] refactor: RecencyScoreFactorTest에서 DB 의존성 제거 및 테스트 리팩토링 #588

- DB 의존성 관련 어노테이션 및 설정 클래스 제거
- 테스트에 불필요한 imports 삭제 및 정리
- `RecencyScoreFactorTest` 초기화 메서드 `setUp` 추가
- Hearit 생성 시 필요한 `createHearit` 메서드 새롭게 정의
- 시간 계산에 따른 점수 감소 테스트 케이스 추가 (`hearitIsCreatedToday`, `returnDecreasedScore`, 등)

Byesol committed:
[BE] refactor: BookmarkScoreFactorTest 리팩토링 및 테스트 케이스 추가 #588

- 테스트 메서드 이름 및 DisplayName 표현 수정으로 가독성 개선
- Bookmark 카테고리 관련 데이터 초기화 코드 간결화
- 북마크 비율 기반 점수 계산 로직 테스트 케이스 개선
- 홓보가 없는 카테고리 및 북마크 없는 사용자의 점수 검증 테스트 추가
- 존재하지 않는 사용자 UUID로 요청 시 예외 발생 테스트 추가

Byesol committed:
[BE] refactor: ExploreScoreTestConfig.java 를 사용하지 않도록 수정 # 588

Byesol committed:
[BE] refactor: GuestExploreScoreProcessorTest의 테스트 표현 개선 및 불필요한 검증 제거 #588

- 테스트 DisplayName 수정으로 가독성 향상
- 비회원 응답에 키워드 정보 포함 여부 테스트 추가
- 불필요한 cursorId 검증 로직 제거

Byesol committed:
[BE] refactor: TestClock 의존 제거 및 insertHearitAt 메서드 추가 #588

- `TestClock` 종속성 제거 및 생성 시간 명시를 위한 `insertHearitAt` 메서드 추가
- 테스트에서 `TestClock.freezeAt/unfreeze` 호출 삭제
- 시간 의존 로직들을 `LocalDateTime` 활용 방식으로 수정
- 관련 테스트 케이스 내부 구현 정리 및 중복 코드 제거

Byesol committed:
[BE] refactor: RecencyScoreFactor 점수 계산 로직 리팩토링 #588

- 상수 이름 `DAYS_PER_POINT`를 `POINT_LOSS_PER_DAY`로 변경하여 의도를 명확히 함
- 점수 계산 로직을 `calculateRecencyScore` 메서드로 분리하여 코드 가독성 개선
- 현재 시간을 계산하기 위한 `LocalDateTime.now()`를 변수로 추출
- 불필요한 `ChronoUnit` import 삭제

Byesol committed:
[BE] refactor: BookmarkScoreFactor 지원 여부 검증 테스트 케이스 추가 #588

- `isSupported` 메서드에 대한 MEMBER 및 GUEST 타입 지원 여부 테스트 케이스 추가
- `UserType` 클래스 import 추가
- 테스트 DisplayName으로 의도 명확히 표현

Byesol committed:
[BE] refactor: ExploreScoreProcessor 내부 메서드 분리 및 호출 구조 개선 #588

- `resolveUserUuid`, `refreshScoresIfNeeded`, `fetchExploreHearits` 메서드 추가 및 책임 분리
- `getExploreHearitsResponse` 메서드에서 위 메서드 호출로 로직 단순화
- `GuestExploreScoreProcessorTest`와 `MemberExploreScoreProcessorTest`에서 새로운 메서드 호출 구조 반영
- 테스트 DisplayName 업데이트로 의도 명확히 표현
- `IllegalStateException` 예외 메시지 수정 (지원하지 않는 요청 → 지원하지 않는 탐색 요청)

Byesol committed:
[BE] refactor: ExploreScoreProcessor의 메서드 시그니처 단순화 및 호출 구조 개선 #588

- `resolveUserUuid` 메서드 제거 및 관련 호출 코드 삭제
- `refreshScoresIfNeeded` 및 `fetchExploreHearits` 메서드의 파라미터 단순화 (`userUuid` 제거)
- `getExploreHearitsResponse`에서 간소화된 메서드 호출 구조 반영
- `GuestExploreScoreProcessorTest` 및 `MemberExploreScoreProcessorTest`의 테스트 코드 업데이트
- 중복된 로직 제거를 통해 코드 가독성과 유지 보수성 개선

Byesol committed:
[BE] refactor: 테스트 코드 주석 추가 및 점수 계산 로직 검증 강화 #588

Byesol committed:
[BE] refactor: Guest/MemberExploreScoreProcessorTest의 테스트 케이스 추가 및 개선 #588

- `isSupported` 메서드에 대한 게스트 및 회원 타입 지원 검증 테스트 추가
- `refreshScoresIfNeeded` 메서드의 점수 저장 로직 검증 테스트 추가
- `fetchExploreHearits` 메서드 결과 반환 테스트 케이스 추가 (데이터 존재 및 미존재 케이스 포함)
- 테스트 DisplayName 수정으로 의도 명확히 표현
- 중복 및 불필요한 로직 제거를 통해 코드 가독성과 유지 보수성 개선

Byesol committed:
[BE] refactor: 테스트 주석 수정 및 점수 계산 대상 Hearit 주석 정리 #588

- 테스트 코드 내 점수 계산 대상 Hearit들의 주석 명확화
- 점수 설명에 불필요한 의문부호 제거로 가독성 개선

Byesol committed:
[BE] refactor: ExploreScoreRefresherTest의 점수 계산 및 검증 로직 개선 #588

- `ExploredHearitQueryRepository` 의존 제거 및 JdbcTemplate 기반 유틸 메서드 추가
- 갱신된 점수의 유효성 검증 로직 추가 (랜덤 스코어 + 최신 점수 계산)
- 테스트 내 Hearit 생성 시간 별 점수 계산 검증으로 로직 신뢰성 강화
- `@MockBean` 어노테이션을 `@MockitoBean`으로 변경
- 불필요한 import 및 중복 코드 제거로 가독성 및 유지 보수성 개선

Byesol committed:
[BE] refactor: ExploreScoreRefresherTest의 갱신 및 점수 검증 테스트 추가 #588

- `refreshScoresWhenCursorIsZeroForGuest` 테스트 추가로 비회원 근거 점수 갱신 로직 검증
- `refreshScoresWhenCursorIsZeroForMember` 테스트 추가로 회원 점수 합산 로직 검증
- 랜덤, 북마크, 최신 스코어 합산 로직 각각에 대한 테스트

Byesol committed:
[BE] refactor: Guest/MemberExploreScoreProcessorTest 테스트 코드 리팩토링 및 DisplayName 수정 #588

- `@MockBean` 어노테이션을 `@MockitoBean`으로 변경하여 테스트 코드를 개선
- DisplayName 수정으로 테스트 의도 명확히 표현
- 중복 코드 제거 및 불필요한 import 삭제로 가독성 및 유지 보수성 개선
- 테스트 메서드 내 사용되지 않는 로직 제거

Byesol committed:
[BE] refactor: ExploreScore 테스트 코드 리팩토링 및 주석 수정 #588

- `@MockBean` 어노테이션을 `@MockitoBean`으로 변경하여 테스트 안정성 개선
- 테스트 내 상수 제거 및 변수 추출로 코드 가독성 향상
- 테스트 주석에서 불필요한 강조 및 의문부호 제거로 표현 개선
- 중복 코드 및 불필요한 import 정리

Byesol committed:
[BE] refactor: ExploreScoreRefresherTest 코드 리팩토링 및 검증 로직 개선 #588

- 중복된 상수 제거, 변수 추출로 코드 가독성 향상
- 각 테스트 내 최신성 점수와 랜덤 점수 계산 로직 명확화
- 불필요한 import 및 Stream API 관련 코드 삭제
- 테스트 DisplayName 수정으로 의도 명확히 표현
- 점수 계산 로직 관련 검증 조건 및 코드 정리

Byesol committed:
[BE] refactor: ExploreScoreRefresherTest의 점수 계산 테스트 코드 리팩토링 #588

- 점수 계산 로직 내 중복 코드를 제거하고 가독성 향상
- 테스트 DisplayName 수정으로 의도를 명확히 표현
- 랜덤, 최신성 점수 계산 테스트에서 예측 점수 변수로 교체
- 불필요한 import 정리 및 코멘트 명확화

Byesol committed:
[BE] refactor: GuestExploreScoreProcessorTest의 테스트 케이스 추가 및 리팩토링 #588

- `refreshScoresIfNeeded` 메서드의 게스트 점수 저장 로직 검증 테스트 추가
- `fetchExploreHearits` 메서드의 점수 순서 응답 검증 테스트 추가
- 점수 데이터가 없을 경우의 빈 리스트 반환 테스트 케이스 업데이트
- 점수 계산 로직 유틸 메서드 추가 (`calculateRecencyScore`, `findExploreScores`)
- 중복 코드 제거 및 불필요한 import 삭제로 코드 가독성 개선
- 테스트 DisplayName 수정으로 테스트 의도 명확히 표현

Byesol committed:
[BE] refactor: MemberExploreScoreProcessorTest의 테스트 케이스 추가 및 리팩토링 #588

- `refreshScoresIfNeededStoresScores` 메서드 구현 및 회원 점수 저장 로직 검증 테스트 추가
- `fetchExploreHearitsReturnsResponses` 메서드의 점수 순서와 북마크 응답 검증 테스트 추가
- 점수 데이터가 없을 경우 빈 리스트 반환 테스트 케이스 추가
-

Byesol committed:
[BE] refactor: ExploreScoreCalculatorTest 리팩토링 및 테스트 케이스 개선 #588

- 게스트 및 회원 점수 계산 테스트의 가독성 향상을 위해 이름 및 변수 정리
- 점수 예상 계산 로직을 별도 변수로 분리하여 코드 중복 제거
- `calculateRecencyScore` 유틸 메서드 추가로 최신성 점수 계산 로직 분리
- `DisplayName` 수정으로 테스트 의도 명확화
- 불필요한 주석 및 import 정리

Byesol committed:
[BE] refactor: HearitExploreServiceTest 리팩토링 및 점수 계산 테스트 로직 개선 #588

- `@MockBean` 어노테이션을 `@MockitoBean`으로 변경하여 테스트 안정성 개선
- 점수 예상 계산 로직 추가로 테스트 일관성 강화
- 중복된 `ExploreScoreProcessor` 인스턴스 제거 및 `@Autowired` 주입 방식으로 변경
- DisplayName 수정 및 주석 정리로 테스트 의도 명확화
- 불필요한 import 및 코드 정리로 가독성 및 유지 보수성 개선

Byesol committed:
[BE] refactor: `nextDouble()` 메서드명을 `getDouble()`로 변경 #588

Byesol committed:
[BE] refactor: `TestClock` 클래스의 `getNow` 메서드 반환 로직 수정 #588

Byesol committed:
[BE] refactor: fetchExploreHearits 메서드명을 getExploreHearits로 변경 #588

Byesol committed:
[BE] refactor: 테스트 코드 내 ExploreScore 생성 로직 통합 및 활용 #588

- `insertMemberExploreScore` 메서드명을 `insertExploreScore`로 변경하여 일관성 유지
- 테스트 코드에서 `jdbcTemplate` 대신 `insertExploreScore` 메서드 사용으로 코드 간소화
- 불필요한 SQL 구문 제거로 코드 가독성 향상 및 유지 보수성 개선
- `GuestExploreScoreProcessorTest` 테스트 코드 내 중복 로직 제거 및 `insertExploreScore` 활용

Byesol committed:
[BE] refactor: GuestExploreScoreProcessorTest에서 GUEST_ID 상수 제거 및 guestUuid 변수로 대체 #588

Byesol committed:
[BE] refactor: IllegalStateException 메시지 수정 #588

Byesol committed:
[BE] refactor: 테스트 코드에서 불필요한 주석 제거 및 assertAll 구조 개선 #588

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/app/explore/application/HearitExploreServiceTest.java:
```diff
@@ -36,135 +39,182 @@
+        ExploreScoreRefresher.class, GuestExploreScoreProcessor.class, MemberExploreScoreProcessor.class})
 @ActiveProfiles("integration-test")
 @AutoConfigureTestDatabase(replace = Replace.NONE)
 class HearitExploreServiceTest {
```
```java
@DataJpaTest
@Sql("/dbclean.sql")
@Import({DbHelper.class, TestJpaAuditingConfig.class, RandomScoreFactor.class, RecencyScoreFactor.class,
        BookmarkScoreFactor.class, ExploreScoreCalculator.class, ExploreScoreCommandRepository.class,
        ExploreScoreRefresher.class, GuestExploreScoreProcessor.class, MemberExploreScoreProcessor.class,
        HearitExploreService.class})
@ActiveProfiles("integration-test")
@AutoConfigureTestDatabase(replace = Replace.NONE)
class HearitExploreServiceTest {

    @MockitoBean
    private RandomNumberGenerator randomNumberGenerator;

    @Autowired
    private DbHelper dbHelper;

    @Autowired
    private HearitExploreService hearitExploreService;
```

Service도 Autowired시킬 수 있는 것 같습니다... 그리고 Repository는 DataJpaTest 사용 시 자동으로 Autowired 되구요

yuyoungrhee replied:
나머지 Service들도 setup에서 직접 조립하지 않고 Autowired 시킬 수 있을 것 같아요
근데 이건 한꺼번에 나머지 테스트들도 정리해야되는 문제라 꼭 이 PR에서 할 필요는 없을 것 같습니다

Byesol replied:
넹 이 부분은 원래 AutoWired로 하긴 했다가 
다른 서비스테스트 컨벤션이랑 일치해야한다고 생각해
'내가 뭘 테스트하고 있는지 알기 위해 setup에서 직접 조립한다.' 
라는의미를 담을 수 있도록 이렇게 하긴 했어요.

Byesol replied:
> 그리고 Repository는 DataJpaTest 사용 시 자동으로 Autowired 되구요


ExploreScoreCommandRepository.class 의 경우만 직접 import 해줬는데 그 이유는 jpaRepository 가 아닌 jdbcRepository기 때문에 자동으로 주입이 안 되어 직접 주입해줘야 합니다!

yuyoungrhee replied:
저도 하다보니까 그전에하던 테스트대로 컨벤션맞춰왔던 거 같은데
지금 생각해보면 굳이 setup에서 조립할 필요 없는 거 같아요
이미 Spring Bean으로 자동주입되는거를 테스트하는 건데 테스트에서만 직접조립하는 게 더 테스트신뢰성이 없는 거 같기도 하고요

그리고 나머지 Repository는 지금 안 쓰이는 거 같아서 삭제해도 될 것 같습니다.

<img width="684" height="639" alt="image" src="https://github.com/user-attachments/assets/4b2274ab-307c-40fe-8021-e0e4f8dee9d7" />

yuyoungrhee replied:
최종 테스트대상 setup vs Autowired는 나중에 다시 논의해봐야될 것 같아서 현재 사용하고있지 않은 코드들만 삭제해주시면 감사하겠습니다! RandomNumebrGenerator도요

Byesol replied:
꼼꼼하게 짚어주셔서 감사합니당. 정리완료했습니다!
RandomNumebrGenerator의 경우 쓰고 있어서 필요합니다!

아래와 같이쓰고 있습니다.
`        given(randomNumberGenerator.getDouble()).willReturn(fixedRandomDouble);`

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/ExploreScoreRefresherTest.java:
```diff
@@ -0,0 +1,215 @@
+                .toDays();
+        double recencyScore = 20.0 - (daysPassed * 0.5);
+        return Math.max(0.0, recencyScore);
+    }
```
Refresher에서 하기엔 너무 과한 테스트 아닐까요?

Refresher 코드만 보기에는 회원/비회원 차이로 기능이 바뀌는 게 아니라 단순히 uuid를 받아 cursorId가 0면 update해주는 그런 로직인 것 같은데요

지금이라면 RecencyScoreFactor에서 쓰는 점수 상수들이 바뀐다면 RefresherTest도 깨지게 되겠네요

점수가 몇 점인지가 아니라, "갱신 작업이 일어났는가"에 대한 결과만 확인해도 충분할 것 같습니다. Mock을 쓰지 않고도 empty에서 갱신됐는지, cursorId는잘부여됐는지 이정도는 검증 가능할 것 같아요.
통합테스트 성격은 이미 HearitExploreService에서 충분히 해주고 있는 것 같고요!

Byesol replied:
저희의 서비스 계층의 테스트들은 항상 통합 테스트로 테스트하는 것이 컨벤션이었어서

이 부분 역시 단위 테스트보다는 통합테스트 형식으로 테스트를 진행했었습니다.

gabean13 commented on backend/app/src/main/java/com/onair/hearit/explore/application/ExploreScoreRefresher.java:
```diff
@@ -0,0 +1,27 @@
+    public void refreshIfNeeded(long cursorId, String userUuid, UserType userType) {
+        if (cursorId != 0L) {
+            return;
+        }
```
이거 !isInitialRequest() 요런식으로 하는거 어떤가용!?
처음 보면 0L인데 왜 return하지? 할 수도 있을 것 같아용

JO-eusan replied:
동의합니다. 그리고 부정문 보다는 아래 같이 하는건 어떨까요?
```
if (cusorId == 0L) {
   // call method
}
```

Byesol replied:
반영했어염

```java
    public void refreshScores(long cursorId, String userUuid, UserType userType) {
        if (isInitialRequest(cursorId)) {
            Map<Long, Double> scores = exploreScoreCalculator.calculateTotalScores(userUuid, userType);
            exploreScoreCommandRepository.insertScores(userUuid, scores);
            exploreScoreCommandRepository.updateCursorIds(userUuid);
        }
    }

    private boolean isInitialRequest(long cursorId) {
        return cursorId == 0L;
    }
```

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/ExploreScoreCalculatorTest.java:
```diff
@@ -0,0 +1,156 @@
+        );
+    }
+
+    private double calculateRecencyScore(LocalDateTime createdAt, LocalDateTime now) {
+        long daysPassed = Duration.between(createdAt.toLocalDate().atStartOfDay(), now.toLocalDate().atStartOfDay())
+                .toDays();
+        double recencyScore = 20.0 - (daysPassed * 0.5);
+        return Math.max(0.0, recencyScore);
+    }
```
CalculatorTest도 각 ScoreFactor의 구현에 강하게 결합되어있어 리팩터링 내성이 약한 테스트가 됩니다.

1. CalculatorTest를 통합테스트로 인정하고 냅두기
2. ScoreFactor가 인터페이스이므로 FakeScoreFactor를 2개정도 만들어 잘 합산되는지 테스트하기

논의가 필요할 것 같습니다.

개인적 의견: HearitExploreServiceTest에서 충분히 탐색서비스에 대한 통합테스트를 수행하고 있는 것 같아서 ScoreFactor에서 오류가 났을 때 깨지는 테스트는 해당 ScoreFactorTest + HearitExploreServiceTest정도면 된다고 생각합니다... ScoreCalculator부터 Refresher까지 깨지면 테스트 신뢰도에 문제가 생길 것 같아서요

yuyoungrhee replied:
Fake를 사용한다면 아래처럼 해볼 수 있겠습니다.

```java
// 테스트에서만 사용하는 FakeScoreFactor 정의
    @TestConfiguration
    static class TestConfig {
        @Bean
        public ScoreFactor alwaysSupportedFactor() {
            return new ScoreFactor() {
                @Override
                public Map<Long, Double> calculate(String uuid, List<Hearit> hearits) {
                    return hearits.stream().collect(Collectors.toMap(Hearit::getId, h -> 100.0));
                }

                @Override
                public boolean isSupported(UserType userType) {
                    return true; // 항상 지원
                }
            };
        }

        @Bean
        public ScoreFactor memberOnlyFactor() {
            return new ScoreFactor() {
                @Override
                public Map<Long, Double> calculate(String uuid, List<Hearit> hearits) {
                    return hearits.stream().collect(Collectors.toMap(Hearit::getId, h -> 10.0));
                }

                @Override
                public boolean isSupported(UserType userType) {
                    return userType == UserType.MEMBER; // 회원만 지원
                }
            };
        }

        @Bean
        public ScoreFactor guestOnlyFactor() {
            return new ScoreFactor() {
                @Override
                public Map<Long, Double> calculate(String uuid, List<Hearit> hearits) {
                    return hearits.stream().collect(Collectors.toMap(Hearit::getId, h -> 1.0));
                }

                @Override
                public boolean isSupported(UserType userType) {
                    return userType == UserType.GUEST; // 게스트만 지원
                }
            };
        }
    }

    @Autowired
    private DbHelper dbHelper;

    @Autowired
    private ExploreScoreCalculator exploreScoreCalculator;

    @DisplayName("사용자 타입에 따라 지원되는 ScoreFactor들을 올바르게 필터링하고 점수를 합산한다")
    @Test
    void correctly_filters_and_sums_scores_based_on_user_type() {
        // given
        Category category = dbHelper.insertCategory(TestFixture.createFixedCategory());
        Hearit hearit = dbHelper.insertHearit(TestFixture.createFixedHearitWith(category));
        String memberUuid = "member-uuid";
        String guestUuid = "guest-uuid";

        // when
        Map<Long, Double> memberScores = exploreScoreCalculator.calculateTotalScores(memberUuid, UserType.MEMBER);
        Map<Long, Double> guestScores = exploreScoreCalculator.calculateTotalScores(guestUuid, UserType.GUEST);

        // then
        assertAll(
                // 회원은 Always(100.0) + MemberOnly(10.0) = 110.0점
                () -> assertThat(memberScores.get(hearit.getId())).isEqualTo(110.0),

                // 게스트는 Always(100.0) + GuestOnly(1.0) = 101.0점
                () -> assertThat(guestScores.get(hearit.getId())).isEqualTo(101.0)
        );
    }
```

Byesol replied:
이렇게 하면

setup 으로 빼두지 않아
 테스트 하나만 읽어도 그 테스트를 이해할 수 있게 만드려는 의도가 퇴색될수 있지 않을까요

Byesol replied:
논의 후 각 테스트에서 동작을 인지할 수 있도록 scoreFactor의 경우 mocking을 쓰는 방식으로 수정했습니다!

gabean13 commented on backend/app/src/test/java/com/onair/hearit/explore/application/HearitExploreServiceTest.java:
```diff
@@ -36,135 +39,182 @@
     }
 
-    @DisplayName("비회원 점수판 생성 - 최신성, 랜덤성")
+    @DisplayName("게스트가 탐색 요청 시, 북마크를 제외한 점수 순으로 정렬하여 반환한다")
```
북마크를 제외했다기 보다 최신, 랜덤 순으로 정렬하여 반환한다 어떰요

Byesol replied:
반영했어염

Byesol replied:
https://github.com/woowacourse-teams/2025-hEARit/pull/615/commits/80a429ef77fef6674b9dadeddc5df566eaa363b1

gabean13 reviewed (commented):
간단 리뷰만 남겼어용 수고했어요 벡터!

yuyoungrhee commented on backend/app/src/test/java/com/onair/hearit/explore/application/scoreprocessor/GuestExploreScoreProcessorTest.java:
```diff
@@ -0,0 +1,224 @@
+                .toDays();
+        double recencyScore = 20.0 - (daysPassed * 0.5);
+        return Math.max(0.0, recencyScore);
+    }
```
GuestExploreScoreProcessor&MemberExploreScoreProcessor 둘다 마찬가지로 Calculator에서 한단계 더 추상화된 ProcessorTest임에도 불구하고 여전히 Factor라는 구현에 결합되어 있네요


Processor가 의존하는 컴포넌트는  ExploreScoreRefresher, ExploredHearitQueryRepository, HearitKeywordRepository 뿐입니다.

개인의견:
ExploreServiceTest에서 이미 service -> 모든하위component -> repository까지 검색하고있는 통합테스트 성격을 띄고 있으므로

각 컴포넌트(Processor, Calculator, Refresher)는 단위테스트 성격으로 가야 맞다고 생각합니다. 어떻게 생각하시나요

Byesol replied:
논의 후 해당 방향으로 수정완료했습다

yuyoungrhee reviewed (changes requested):
Service Autowired 관련 + 단위테스트 성격 관련 리뷰 남겼습니다!

JO-eusan commented on backend/app/src/main/java/com/onair/hearit/explore/application/scoreprocessor/ExploreScoreProcessor.java:
```diff
@@ -1,12 +1,14 @@
     boolean isSupported(UserInfo userInfo);
 
-    List<ExploredHearitResponse> getExploreHearitsResponse(UserInfo userInfo, long cursorId, int size);
+    void refreshScoresIfNeeded(UserInfo userInfo, long cursorId);
```
`refreshScores` 만 해도 충분한 네이밍이라고 생각합니다. `IfNeeded`까지 외부에서 알아야할까요?

Byesol replied:
수정했어염
https://github.com/woowacourse-teams/2025-hEARit/pull/615/commits/e4f54ac9520293389f26a8499a686aa6e04d55ec

JO-eusan reviewed (changes requested):
벡터 고생하셨어요!
테스트 리뷰는 많이 받으신거 같아서 프로덕션 코드에서 몇가지 리뷰 남겨두었습니다.

Byesol committed:
[BE] refactor: ExploreScoreCalculatorTest 리팩토링 및 모킹된 ScoreFactor 테스트 추가 #588

Byesol committed:
[BE] refactor: ExploreScoreRefresherTest 리팩토링 및 불필요한 모킹 제거 #588

Byesol committed:
[BE] refactor: 테스트 코드 내 불필요한 코드와 주석 제거 #588

- `MemberExploreScoreProcessorTest` 내 불필요한 주석 제거 및 가독성 개선.
- `GuestExploreScoreProcessorTest`의 중복 코드 제거 및 `insertExploreScore` 메서드 활용.
- 테스트 내 `ExploreScore` 생성 방식을 통합하여 중복 로직 최소화.
- import 최적화 및 유지 보수성 향상.

Byesol committed:
[BE] refactor: 탐색 서비스 테스트의 정렬 로직 관련 설명 수정 #588

Byesol committed:
[BE] refactor: ExploreScoreRefresher에 초기 요청 판단 메서드 분리 #588

Byesol committed:
[BE] refactor: refreshIfNeeded 메서드명을 refreshScores로 변경 #588

Byesol committed:
[BE] refactor: refreshScores 메서드 내 초기 요청 조건문 리팩토링 #588

Byesol committed:
[BE] refactor: 테스트 코드에서 불필요한 Autowired 제거 #588

Byesol committed:
[BE] Merge remote-tracking branch 'origin/develop-be' into refactor/#588-explore-test

Byesol committed:
[BE] refactor: 병합하며 생긴 문제 해결 #588

Byesol committed:
[BE] refactor: 병합하며 생긴 문제 해결 #588

Byesol committed:
[BE] refactor: ExploreScoreCalculatorTest 파일에 패키지 선언 추가 #588

Byesol committed:
[BE] refactor: ExploreScoreCalculatorTest 파일에서 불필요한 개행 제거 #588

gabean13 reviewed (approved):
벡터 고생했어욧!!!

yuyoungrhee reviewed (approved):
반영 모두 잘 확인했습니다! 고생 많으셨어요 벡터~

JO-eusan reviewed (approved):
벡터 연휴에도 고생하셨어요~ Approve할게욤!
